### PR TITLE
Make `outputHashAlgo` accept `"nar"`, stay in sync

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Now that we have a few things identifying content address methods by name, we should be consistent about it.

Move up the `parseHashAlgoOpt` for tidiness too.

# Context

Discussed this change for consistency's sake as part of #8876

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: doc/manual/src/language/advanced-attributes.md
Comment (edolstra) [inline doc/manual/src/language/advanced-attributes.md:206]:
```suggestion
      - `"recursive"` or `"nar"`\
```

We may want to say that `recursive` is preferred because it's backward compatible. Indeed the downside of introducing such aliases is that they're not forward compatible.

Comment (Ericson2314) [reply doc/manual/src/language/advanced-attributes.md:206]: How is the note that I added?

I am not sure "recursive" is preferred. If anything I would hope it would be deprecated and people use "nar" for consistent and explicitness. But that is not a priority for me so I left not either being "preferred", just stating the facts so users can make their own choice.

File: src/libexpr/primops.cc
Comment (thufschmitt) [inline src/libexpr/primops.cc:1125]: Don't we instead want to accept `recursive as a file ingestion method for consistency?

Comment (Ericson2314) [reply src/libexpr/primops.cc:1125]: I am not sure what you mean, I think change `ContentAddressMethod::parse` so this one is accepted everywhere (e.g. the new CLI too)?

Comment (thufschmitt) [reply src/libexpr/primops.cc:1125]: Yes, that's what I meant

Comment (Ericson2314) [reply src/libexpr/primops.cc:1125]: I forget whether we discussed this before, but I do want to over time steer people to the new name (which is clearer) in the new CLI.

Comment (Ericson2314) [reply src/libexpr/primops.cc:1125]: (Also with git support, which I hope to mere very soon, NAR isn't the only "recursive" format anyways.)

File: doc/manual/src/language/advanced-attributes.md
Comment (edolstra) [reply doc/manual/src/language/advanced-attributes.md:206]: Using `nar` instead of `recursive` makes a Nix expression unnecessarily backwards-incompatible. E.g. if somebody adds a package that uses `nar` to Nixpkgs, it won't work on older versions of Nix while it otherwise probably would have. That's the danger of adding aliases like this.

Comment (roberth) [reply doc/manual/src/language/advanced-attributes.md:206]: If we do it, I think it should warn loudly for at least 2 years and until all implementations support it, whichever is later.
If John wants to manage such a long term thing, I think we could let him. Before you know it, it'll be 2026 and time will have solved the problem.

Comment (Ericson2314) [reply doc/manual/src/language/advanced-attributes.md:206]: Yeah at the end of the day I am OK with warnings, docs saying don't use this, etc. The one thing I care about is that this also support `"nar"` so we could conceivably, many years from now, get the ecosystem off `"recursive"`.

Comment (Ericson2314) [reply doc/manual/src/language/advanced-attributes.md]: (I applied the suggestion and also reordered the explanations below.)

**Review by thufschmitt — APPROVED** [review]

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-02-28.
Approved, modulo the test failure and the merge conflicts

<details>

- New aliases are annoying because cause backwards-compat issues
    - Acceptable trade of

**Comment by roberth** [comment]
This would be a case where a release note becomes more valuable over time.
@Ericson2314 could you add one? (with a recommendation not to use it yet)

**Pull request merged**
